### ISSUE TITLE
Ruff: 0.4.2 generated invalid fix it code

### ISSUE THREAD

**Issue by Skylion007** [astral-sh/ruff#11166]
```python
                msg="\n".join(
                    [
                        "alpha = alpha_c + %.2g" % shift,
                        "expected_grad: %.5g" % expected_grad,
                        "actual_grad: %.5g" % actual_grad,
                        "error = %.2g" % torch.abs(expected_grad - actual_grad).max(),
                    ]
```
became
```
                        f"alpha = alpha_c + {shift:.2g}",
                        f"expected_grad: {expected_grad:.5g}",
                        f"actual_grad: {actual_grad:.5g}",
                        f"error = {torch.abs(expected_grad - actual_grad).max():.2g}",
 ```
which throws an error. I know it's an unsafe fix, but unsafe fixes should not generated invalid code. ;-;

**Comment by AlexWaygood** [comment]
Guessing this is due to [the recent change to UP031](https://github.com/astral-sh/ruff/pull/11019)? Cc. @plredmond

**Comment by charliermarsh** [comment]
Why can't tensors be passed-in there, but can be formatted with `%` formatting?

**Comment by Skylion007** [comment]
@charliermarsh I think it's related to classes overriding the `__format__` magic method. Does the '%' method cast it to a string before calling `__format__` on it or something?

**Comment by Skylion007** [comment]
@charliermarsh Ah, apparently in the first code, it's implicitly converted before: "Here, you're using the old-style string formatting (%-formatting), where shift, a PyTorch tensor, is implicitly converted to a numeric type (float or integer) when it's formatted. This conversion happens before the formatting is applied, so the process doesn't raise any issues."

```python
import torch
x = torch.tensor([1.0])
print("Value: %.2f" % x)  # Works because x is cast as a float before __format__ is called
print(f"Value: f{x:.5g}") # fails because Tensor.__format__ is called
```

**Comment by Skylion007** [comment]
I think explicitly casting to float in all float specific `__format__` string fixits could solve this issue?

**Comment by charliermarsh** [comment]
I think that makes sense. E.g., when there's a `:.5g` specifier?

**Comment by Skylion007** [comment]
There are probably other specifies that do this casting f (or maybe some to int for d) etc.

**Comment by plredmond** [comment]
@charliermarsh perhaps we could add this kind of explicit-conversion to the UP031 rule after we can infer the types of the expressions? alternatively, since we parse the format string we could use that information to select an explicit-conversion function?
